doc/md/fundamentals/2-types/10-options.md

```diff
@@ -0,0 +1,227 @@
+
+The same logic can be expressed using a `switch`, though the result is  more verbose and introduces an additional level of nesting:
+
+``` motoko
+func get<T>(option : ?T, defaultValue : T) : T {
+  switch option {
+    case null defaultValue;
+    case (?value) value;
+  }
+};
+```
+
+:::tip
+Although convenient for option patterns, `let-else` also works with other types of pattern,
+handling pattern-match failure using the `else` branch.
+:::
+
+### Applying transformations to options
+
+The `Option.map` function applies a transformation only if the value is present.
+
+```motoko
+import Option "mo:base/Option";
+
+let number : ?Nat = ?10;
+Option.map<Nat, Nat>(number, func(x : Nat) = x * 2); // ?20
+```
+
+In this example, if `number` is `null`, `map` ensures the result remains `null` instead of performing an invalid operation.
+
+### Applying an optional function
+
+Sometimes, both the function and value are optional. `Option.apply` calls a function only if both are present. This is useful when chaining optional operations that may return `null`.
+
+```motoko
+import Option "mo:base/Option";
+
+let maybeIncrement : ?(Nat -> Nat) = ?(func x: Nat = x + 1);
+let maybeValue : ?Nat = ?10;
+
+Option.apply<Nat, Nat>(maybeValue, maybeIncrement); // ?11
+```
+
+If either `maybeFunction` or `maybeValue` is `null`, the result remains `null`.
+
+### Combining multiple optional values
+
+When working with multiple optional values, using `Option.chain` processes them safely without unnecessary `switch` statements.
+
+```motoko
+import Option "mo:base/Option";
+
+let firstName : ?Text = ?"Motoko";
+let lastName : ?Text = ?"Ghost";
+
+func combineNames(f : Text, l : Text) : Text {
+    f # " " # l;
+};
+
+Option.chain<Text, Text>(firstName, func (first : Text) {
+    Option.map<Text, Text>(lastName, func(last : Text) { combineNames(first, last) });
+});
+
+// ?("Motoko Ghost")
+```
+
+If either `firstName` or `lastName` is `null`, the result remains `null`.
+
+## Option blocks (`do ?`)
+
+Option blocks use the syntax `do ? <block>` to handle optional values of type `?T` without needing nested switch statements. It produces a value of type `?T`, when `<block>` has type `T` and, importantly, introduces the possibility of a break from `<block>.`
+
+Within a `do ? <block>`, the `null` break `<exp> !` tests whether the result of the expression, `<exp>` of unrelated option type, `?U`, is `null`.
+
+If the result is `null`, control immediately exits the `do ? <block>` with value `null`. Otherwise, the result of `<exp>` must be an option value `?v`, and evaluation of `<exp> !` proceeds with its contents, `v` of type `U`.
+
+The `do ? <block>` is similar to how the `?` operator works in languages like Rust.
+
+```motoko no-repl
+// Introduces an option block that returns a value of type ?T
+let result: ?T = do ? {
+  // The ! operator (null break) unwraps optional values inside the block
+  someOptionalValue!  // unwraps the value or short-circuits
+}
+```
```

Comment:
Your example isn't wrong, but also doesn't do anything very interesting apart from unwrap and re-wrap an option.
Maybe this is better?
```suggestion
```motoko 
 // Returns the sum of optional values `x` and `y` or `null`, if either is `null`
func addOpt(x:?Nat, y:?Nat) : ?Nat {
    do ? { 
      x! + y!  
    }
}
```